File: frontend/src/types/inventario.ts
export interface InventarioVivero {
  id_inventario: number;
  codigo_lote: string;
  id_variedad?: number | null;
  id_portainjerto?: number | null;
  id_vivero?: number | null;
  id_especie?: number | null;
  id_pmg?: number | null;
  id_bodega?: number | null;
  tipo_planta?: string | null;
  tipo_injertacion?: string | null;
  tipo_patron?: string | null;
  ubicacion?: string | null;
  cantidad_inicial: number;
  cantidad_actual: number;
  cantidad_minima: number;
  cantidad_comprometida?: number;
  fecha_ingreso: string;
  ano_plantacion?: number | null;
  origen?: string | null;
  estado: string;
  observaciones?: string | null;
  fecha_creacion?: string;
  fecha_modificacion?: string | null;
}

export interface MovimientoInventario {
  id_movimiento: number;
  id_inventario: number;
  id_planta?: number | null;
  tipo: string;
  cantidad: number;
  saldo_anterior?: number | null;
  saldo_nuevo?: number | null;
  motivo?: string | null;
  referencia_destino?: string | null;
  usuario?: string | null;
  fecha_movimiento: string;
}

export interface InventarioStats {
  total_lotes: number;
  total_stock: number;
  lotes_disponibles: number;
  lotes_agotados: number;
}

export interface GuiaDespacho {
  id_guia: number;
  numero_guia?: string | null;
  id_bodega_origen?: number | null;
  id_testblock_destino?: number | null;
  estado: string;
  total_plantas?: number | null;
  responsable?: string | null;
  motivo?: string | null;
  usuario?: string | null;
  fecha_creacion: string;
}

export interface InventarioTestBlock {
  id_inventario_tb: number;
  id_inventario: number;
  id_cuartel: number;
  codigo_lote: string | null;
  variedad: string | null;
  portainjerto: string | null;
  cantidad_asignada: number;
  cantidad_plantada: number;
  disponible: number;
  estado: string;
  fecha_despacho: string | null;
  fecha_completado: string | null;
  observaciones: string | null;
}

File: backend/app/models/inventario.py
"""Inventario vivero, movimientos, guias despacho, inventario_testblock models."""

from datetime import datetime, date
from typing import Optional
from sqlmodel import SQLModel, Field, Column
import sqlalchemy as sa


class InventarioVivero(SQLModel, table=True):
    __tablename__ = "inventario_vivero"
    id_inventario: Optional[int] = Field(default=None, primary_key=True)
    codigo_lote: str = Field(sa_column=Column(sa.String(50), unique=True, nullable=False))
    id_variedad: Optional[int] = Field(default=None, foreign_key="variedades.id_variedad")
    id_portainjerto: Optional[int] = Field(default=None, foreign_key="portainjertos.id_portainjerto")
    id_vivero: Optional[int] = Field(default=None, foreign_key="viveros.id_vivero")
    id_especie: Optional[int] = Field(default=None, foreign_key="especies.id_especie")
    id_pmg: Optional[int] = Field(default=None, foreign_key="pmg.id_pmg")
    id_bodega: Optional[int] = Field(default=None, foreign_key="bodegas.id_bodega")
    tipo_planta: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(50)))
    tipo_injertacion: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(50)))
    tipo_patron: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(50)))
    ubicacion: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(200)))
    cantidad_inicial: int
    cantidad_actual: int
    cantidad_minima: int = Field(default=0)
    cantidad_comprometida: int = Field(default=0)
    fecha_ingreso: date
    ano_plantacion: Optional[int] = Field(default=None)
    origen: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(100)))
    estado: Optional[str] = Field(default="disponible", sa_column=Column(sa.String(20)))
    observaciones: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(None)))
    fecha_creacion: Optional[datetime] = Field(default_factory=datetime.utcnow)
    fecha_modificacion: Optional[datetime] = Field(default=None)


class MovimientoInventario(SQLModel, table=True):
    __tablename__ = "movimientos_inventario"
    id_movimiento: Optional[int] = Field(default=None, primary_key=True)
    id_inventario: Optional[int] = Field(default=None, foreign_key="inventario_vivero.id_inventario")
    id_planta: Optional[int] = Field(default=None, foreign_key="plantas.id_planta")
    tipo: str = Field(sa_column=Column(sa.String(30), nullable=False))
    cantidad: int
    saldo_anterior: Optional[int] = Field(default=None)
    saldo_nuevo: Optional[int] = Field(default=None)
    motivo: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(200)))
    referencia_destino: Optional[str] = Field(default=None, sa_column=Column(sa.String(50)))
    id_evento: Optional[int] = Field(default=None)
    usuario: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(50)))
    fecha_movimiento: Optional[datetime] = Field(default_factory=datetime.utcnow)


class InventarioTestBlock(SQLModel, table=True):
    __tablename__ = "inventario_testblock"
    id_inventario_tb: Optional[int] = Field(default=None, primary_key=True)
    id_inventario: Optional[int] = Field(default=None, foreign_key="inventario_vivero.id_inventario")
    id_cuartel: Optional[int] = Field(default=None, foreign_key="cuarteles.id_cuartel")
    cantidad_asignada: Optional[int] = Field(default=None)
    cantidad_plantada: int = Field(default=0)
    estado: Optional[str] = Field(default="pendiente", sa_column=Column(sa.String(20)))
    fecha_despacho: Optional[date] = Field(default=None)
    fecha_completado: Optional[date] = Field(default=None)
    observaciones: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(None)))
    usuario_creacion: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(50)))
    fecha_creacion: Optional[datetime] = Field(default_factory=datetime.utcnow)


class GuiaDespacho(SQLModel, table=True):
    __tablename__ = "guias_despacho"
    id_guia: Optional[int] = Field(default=None, primary_key=True)
    numero_guia: str = Field(sa_column=Column(sa.String(20), nullable=False))
    id_bodega_origen: int = Field(foreign_key="bodegas.id_bodega")
    id_testblock_destino: int = Field(foreign_key="testblocks.id_testblock")
    estado: Optional[str] = Field(default="pendiente", sa_column=Column(sa.String(20)))
    total_plantas: int
    responsable: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(100)))
    motivo: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(500)))
    usuario: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(50)))
    fecha_creacion: Optional[datetime] = Field(default_factory=datetime.utcnow)
    fecha_despacho: Optional[datetime] = Field(default=None)
    fecha_recepcion: Optional[datetime] = Field(default=None)
    notas: Optional[str] = Field(default=None, sa_column=Column(sa.NVARCHAR(None)))
    activo: Optional[bool] = Field(default=True)
